src/common_subset.rs

```diff
@@ -116,57 +117,117 @@ impl<NodeUid: Clone + Debug + Display + Eq + Hash + Ord> CommonSubset<NodeUid> {
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
+        // The result defaults to error.
+        let mut result = Err(Error::NoSuchAgreementInstance);
+
+        // Send the message to the local instance of Agreement
+        if let Some(mut agreement_instance) = self.agreement_instances.get_mut(&element_proposer_id)
+        {
+            // Optional output of agreement and outgoing agreement
+            // messages to remote nodes.
+            result = if agreement_instance.terminated() {
+                // This instance has terminated and does not accept input.
+                Ok((None, VecDeque::new()))
+            } else {
+                // Send the message to the agreement instance.
+                agreement_instance
+                    .on_input(msg_sender_id, &amessage)
+                    .map_err(Error::from)
+            }
+        }
 
-                Err(Error::NotImplemented)
+        if let Ok((output, mut outgoing)) = result {
+            // Process Agreement outputs.
+            if let Some(b) = output {
+                outgoing.append(&mut self.on_agreement_result(element_proposer_id, b));
             }
+
+            // Check whether Agreement has completed.
+            Ok((
+                self.try_agreement_completion(),
+                outgoing.into_iter().map(Output::Agreement).collect(),
+            ))
+        } else {
+            // error
+            result
+                .map(|(_, messages)| (None, messages.into_iter().map(Output::Agreement).collect()))
         }
     }
 
     /// Callback to be invoked on receipt of a returned value of the Agreement
     /// instance `uid`.
-    ///
-    /// FIXME: It is likely that only one `AgreementMessage` is required because
-    /// Figure 11 does not count the number of messages but the number of nodes
-    /// that sent messages.
-    fn on_agreement_result(&mut self, uid: NodeUid, result: bool) -> VecDeque<AgreementMessage> {
+    fn on_agreement_result(
+        &mut self,
+        element_proposer_id: NodeUid,
+        result: bool,
+    ) -> VecDeque<AgreementMessage> {
         let mut outgoing = VecDeque::new();
         // Upon delivery of value 1 from at least N − f instances of BA, provide
         // input 0 to each instance of BA that has not yet been provided input.
         if result {
-            self.agreement_true_outputs.insert(uid);
+            self.agreement_results.insert(element_proposer_id, result);
+            // The number of instances of BA that output 1.
+            let results1 = self.agreement_results.values().filter(|v| **v).count();
 
-            if self.agreement_true_outputs.len() >= self.num_nodes - self.num_faulty_nodes {
-                let instances = &mut self.agreement_instances;
-                for (_uid0, instance) in instances.iter_mut() {
+            if results1 >= self.num_nodes - self.num_faulty_nodes {
+                for instance in self.agreement_instances.values_mut() {
```

Comment:
See above: I'm worried that this could break the agreement if the instance has already progressed beyond the first round. (But it's not quite clear in the paper and I haven't thought much about it yet.)

Reply:
In the paper, BA instances can set the estimated value only for the next epoch. So, if an instance hasn't been provided with input `true` from the ACS instance, the estimated value of the epoch 0 will remain empty until the ACS instance fills it in using `Agreement::set_input(false)`. Then that BA can terminate as soon as the coin value is `false` due to the termination criterion.